scripts/cleanup-e2e-codebuild-resources.ts

```diff
@@ -0,0 +1,1107 @@
+const STALE_DURATION_MS = 2 * 60 * 60 * 1000; // 2 hours in milliseconds
+
+/*
+ * Exit on expired token as all future requests will fail.
+ */
+const handleExpiredTokenException = (): void => {
+  console.log('Token expired. Exiting...');
+  process.exit();
+};
+
+/**
+ * Check if a resource is stale based on its created date
+ * @param created
+ * @returns
+ */
+const isStale = (created: string | Date | undefined): boolean => {
+  let normalizedDate;
+  if (typeof created === 'string') {
+    normalizedDate = new Date(created);
+  } else {
+    normalizedDate = created;
+  }
+  const now = new Date().getTime();
+  const isStale = normalizedDate ? now - normalizedDate.getTime() > STALE_DURATION_MS : false;
+  return isStale;
+};
+
+/**
+ * We define a resource as viable for deletion if it matches TEST_REGEX in the name, and if it is > STALE_DURATION_MS old.
+ */
+const testBucketStalenessFilter = (resource: S3.Bucket): boolean => {
+  const isTestResource = !!resource?.Name?.match(BUCKET_TEST_REGEX);
+  return isTestResource && isStale(resource.CreationDate);
+};
+
+const testRoleStalenessFilter = (resource: IAM.Role): boolean => {
+  const isTestResource = !!resource?.RoleName?.match(IAM_TEST_REGEX);
+  return isTestResource && isStale(resource.CreateDate);
+};
+
+const testUserPoolStalenessFilter = (resource: CognitoIdentityServiceProvider.UserPoolDescriptionType): boolean => {
+  const isTestResource = !!resource?.Name?.match(USER_POOL_TEST_REGEX);
+  return isTestResource && isStale(resource.CreationDate);
+};
+
+const testAppSyncApiStalenessFilter = (resource: AppSync.GraphqlApi): boolean => {
+  const isTestResource = !!resource?.name?.match(APPSYNC_TEST_REGEX);
+  const createTimeTagValue = resource?.tags?.['codebuild:create_time'];
+  let isStaleResource = true;
+  if (createTimeTagValue) {
+    const createTime = new Date(createTimeTagValue);
+    isStaleResource = isStale(createTime);
+  }
+  return isTestResource && isStaleResource;
+};
+
+const testPinpointAppStalenessFilter = (resource: Pinpoint.ApplicationResponse): boolean => {
+  const isTestResource = !!(resource.Name.match(PINPOINT_TEST_REGEX) && resource.CreationDate);
+  return isTestResource && isStale(resource.CreationDate);
+};
+
+/**
+ * Get all S3 buckets in the account, and filter down to the ones we consider stale.
+ */
+const getOrphanS3TestBuckets = async (account: AWSAccountInfo): Promise<S3BucketInfo[]> => {
+  const s3Client = new S3(getAWSConfig(account));
+  const listBucketResponse = await s3Client.listBuckets().promise();
+  const staleBuckets = listBucketResponse?.Buckets?.filter(testBucketStalenessFilter);
+  return staleBuckets?.map((it) => ({ name: it.Name, createTime: it.CreationDate })) ?? [];
+};
+
+/**
+ * Get all iam roles in the account, and filter down to the ones we consider stale.
+ */
+const getOrphanTestIamRoles = async (account: AWSAccountInfo): Promise<IamRoleInfo[]> => {
+  const iamClient = new IAM(getAWSConfig(account));
```

Comment:
I might be missing something, as I didn't implement the original script (cc @lazpavel), but I don't like that as much because we will have to update the global config for each account. Introducing global state makes the implementation less flexible and reliant on side effects imho.

Reply:
Ah yeah if we are switching accounts as the script runs then this is okay